cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java

```diff
@@ -1010,21 +1013,30 @@ private QueryBuilder createMainQuery(SearchRequest request) {
     // all terms must occur in the text field as selected by the scope in the search request.
     String textField = request.getScope() == null ? TEXT_FIELD : request.getScope().name().toLowerCase();
 
-    // split the query into its terms and iterate over all terms
-    Iterable<String> terms = Splitter.on(SPACE_SEPARATOR_PATTERN)
-      .omitEmptyStrings().trimResults().split(request.getQuery());
-    List<QueryBuilder> termQueries = new ArrayList<>();
-    for (String term : terms) {
-      termQueries.add(createTermQuery(term, textField, request));
+    // split the query into its parsed terms and create corresponding QueryBuilders for each
+    Map<QueryBuilder, QueryTerm> termQueries = new HashMap<>();
+    for (QueryTerm queryTerm : QueryParser.parse(request.getQuery())) {
+      termQueries.put(createTermQuery(queryTerm.getTerm(), textField, request), queryTerm);
     }
     if (termQueries.isEmpty()) {
       return QueryBuilders.matchAllQuery();
     }
-    if (termQueries.size() == 1) {
-      return termQueries.get(0);
+    //populate boolQuery clauses based on corresponding queryTerm qualifier
+    BoolQueryBuilder boolQuery = QueryBuilders.boolQuery();
+    for (QueryBuilder queryBuilder : termQueries.keySet()) {
+      if (termQueries.get(queryBuilder).getQualifier() == Qualifier.REQUIRED) {
+        boolQuery.must(queryBuilder);
+      } else {
+        boolQuery.should(queryBuilder);
+      }
+    }
+
+    //if optional terms are the only terms available, require at least one
+    if (!boolQuery.should().isEmpty() && boolQuery.must().isEmpty()) {
+      boolQuery = boolQuery.minimumShouldMatch(1);
```

Comment:
what happens if you always set this to 1. That is, if there is one or more required term, will the minimumShouldMatch(1) require that one of the optional terms has to be there? Do you explicitly have to set it to 0?